### ISSUE TITLE
Differences with Bahdanau et al., 2014

### ISSUE THREAD

**Issue by marcobrando** [datalogue/keras-attention#4]
Thank you for sharing your code!

I spotted some minor differences between your implementation and the attention mechanism described in Bahdanau, Dzmitry, Kyunghyun Cho, and Yoshua Bengio, "Neural machine translation by jointly learning to align and translate.", [arXiv preprint arXiv:1409.0473] (2014) (https://arxiv.org/pdf/1409.0473.pdf).

I) In the initialization part, you have `s0 = activations.tanh(K.dot(inputs[:, 0], self.W_s))`.
If I'm not wrong, `inputs[:, 0]` should correspond to h_forward_1, whilst in the reference paper should be h_backward_1

II) In the output generation step you have `yt = activations.softmax(K.dot(ytm, self.W_o)+ K.dot(stm, self.U_o)+ K.dot(context, self.C_o)+ self.b_o)` whilst in the paper it's a little bit more complicated, see section A.2.2.  

Thank you again for sharing this attention implementation.

**Comment by zafarali** [comment]
I'm glad you enjoyed it! Yes, there are some minor differences that were made to keep the tutorial and code descriptive.

> I) In the initialization part, you have `s0 = activations.tanh(K.dot(inputs[:, 0], self.W_s))`.
If I'm not wrong, `inputs[:, 0]` should correspond to h_forward_1, whilst in the reference paper should be h_backward_1

here `inputs` is the input to the AttentionLayer and are the _annotations_ returned by the BLSTM. In the paper these are called `h = [h_forward_1; h_backward_1]` (Equation 7 - concatenation of forward and backward). This is fed into the AttentionLayer and is used to calculate `s0` (last paragraph of page 13). Therefore, `inputs` and `h` are the same thing 😄 

You are right in that the naming of the variables could have been made more consistent!

> II) In the output generation step you have `yt = activations.softmax(K.dot(ytm, self.W_o)+ K.dot(stm, self.U_o)+ K.dot(context, self.C_o)+ self.b_o)` whilst in the paper it's a little bit more complicated, see section A.2.2.  

Good observation! As mentioned in point 4 of the "Equations" section of the blog, this was done to keep things simple. As the Max Out layer is not - to my knowledge - a common layer, I replaced it with a simple fully connected layer. 

Hope these clear up your confusions and let me know if you have any more questions.
